### PR TITLE
Refactor BSpan runtime API

### PR DESCRIPTION
**Pull request by nadundesilva**
## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java [hunk 1]
```diff
@@ -15,151 +15,116 @@
...
-     * Active Ballerina {@link ObserverContext}.
-     */
-    private final ObserverContext observerContext;
+    public static BSpan start(String serviceName, String operationName, boolean isClient) {
+        return new BSpan(null, serviceName, operationName, isClient);
+    }
+
     /**
-     * Open tracer specific span.
+     * Start a new with a parent using parent span.
```
Comment (tharinduwijewardane) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java:74]:
```suggestion
     * Start a new span with a parent using parent span.
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java [hunk 2]
```diff
@@ -15,151 +15,116 @@
...
-                ? TraceConstants.TAG_SPAN_KIND_CLIENT
-                : TraceConstants.TAG_SPAN_KIND_SERVER);
+    public static BSpan start(BSpan parentSpan, String serviceName, String operationName, boolean isClient) {
+        return new BSpan(parentSpan.span.context(), serviceName, operationName, isClient);
     }
 
-    public void startSpan() {
-        manager.startSpan(getParentBSpan(), this);
+    /**
+     * Start a new with a parent using parent trace context.
```
Comment (tharinduwijewardane) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java:87]:
```suggestion
     * Start a new span with a parent using parent trace context.
```

Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java]: Fixed in fc0e0f80fdfa26ccbb5cdeb99deac877c0bb8261

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java (hunk 1 above)
Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java]: Fixed in fc0e0f80fdfa26ccbb5cdeb99deac877c0bb8261

**Comment by nadundesilva** [comment]
Synced with master to get the latest changes and verify build checks.

**Comment by nadundesilva** [comment]
The performance comparison between the master branch and the changes are as follows.
perf_run
The tests were conducted on a single node AKS cluster with ContainerSupport enabled in the JVM.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java [hunk 3]
```diff
@@ -15,151 +15,116 @@
...
  * @since 0.964.1
  */
 public class BSpan {
-
-    private static final TraceManager manager = TraceManager.getInstance();
+    private final Tracer tracer;
+    private final Span span;
+
+    private BSpan(SpanContext parentSpanContext, String serviceName, String operationName, boolean isClientContext) {
+        tracer = TracersStore.getInstance().getTracer(serviceName);
```
Comment (manuranga) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java:41]: shall we remove the duplicity by moving below constructor and this ones logic to the static method, and make the constructors dumber.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/TracingUtils.java [hunk 4]
```diff
@@ -47,24 +46,25 @@ private TracingUtils() {
...
-        BSpan span = new BSpan(observerContext, isClient);
-        span.setServiceName(observerContext.getServiceName());
-
-        span.setOperationName(observerContext.getOperationName());
-        if (isClient) {
-            observerContext.addProperty(PROPERTY_TRACE_PROPERTIES, span.getProperties());
+        BSpan span;
+        if (observerContext.getParent() != null) {
+            BSpan parentSpan = observerContext.getParent().getSpan();
+            span = BSpan.start(parentSpan, observerContext.getServiceName(), observerContext.getOperationName(),
```
Comment (manuranga) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/TracingUtils.java:52]: lets move `observerContext.getServiceName()` and `observerContext.getOperationName()` to variables.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java (hunk 3 above)
Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/BSpan.java]: Fixed in 7795f8a

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/TracingUtils.java (hunk 4 above)
Comment (nadundesilva) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/observability/tracer/TracingUtils.java]: Fixed in 6912281

**Review by manuranga — APPROVED** [review]

**Pull request merged**